check-style.sh

```diff
@@ -0,0 +1,136 @@
+# To save time, we only run over the files in the current commit in the
+# precommit hook. We then run the more extensive full check as part of our
+# branch protection rules. For a fuller discussion, please see
+# https://github.com/reboot-dev/respect/issues/1374.
+#
+print_help() {
+    echo "Use as $0 [--full|--pre-commit]"
+    echo ""
+    echo " --full to check all files under version control."
+    echo " --pre-commit to check the files in the current commit."
+    echo ""
+}
+case $1 in
+    "--full" )
+        mode="full" # Used later to decide how to invoke clang format.
+        # Get all files under version control. These are the superset of files
+        # we will want to check.
+        affected_files=$(git ls-tree --full-tree --name-only -r HEAD)
+        ;;
+    "--pre-commit" )
+        mode="pre-commit" # Used later to decide how to invoke clang format.
+        # Get all files in current commit. We will restrict the checking to
+        # these.
+        affected_files=$(git diff --cached --name-only --diff-filter=ACM HEAD)
+        ;;
+    * )
+        echo "Unknown argument: $1"
+        print_help
+        exit 2
+esac
+
+
+# Unset variable would be a sign of programmer error. We are not using '-e' in
+# this script as we'd like to handle these cases ourselves where relevant, i.e.,
+# allow more than one code check failure per run.
+set -u
+
+# Work from the git root. This is important to help some tools pick up the
+# correct configuration.
+git_root=$(git rev-parse --show-toplevel)
+cd "${git_root}"
+
+# Define a cummulative status code for the script. The value is Updated through
+# 'run_check' when running checks. A status code creater than 0 will indicate
+# that one or more checks failed.
+status_code=0
+
+# Run a code check command and update the cummulative error code.
+run_check() {
+    $@
+    status_code=$(($status_code + $?))
+}
+
+# Helper function to filter affected files by extension.
+get_files_by_extension() {
+    filter=""
+    for ext in $@; do
+        # We want a `.` to be a full stop, not a regexp wildcard.
+        filter="$(echo $ext | sed -e 's|^\.|\\.|')$|${filter}"
+    done
+    filter=$(echo $filter | sed -e 's/|$//')
+
+    echo $affected_files | tr ' ' '\n' | egrep $filter
+}
+
+# Check files that we can clang-format.
+clang_format_files=$(get_files_by_extension .cc .cpp .h .hpp .proto)
+if [ ! -z "${clang_format_files}" ]; then
+    # Run clang-format. In fact, run our wrapper script around-clang format that
+    # also contains logic for checking the line length.
+    #
+    # ISSUE https://github.com/reboot-dev/respect/issues/1371: This script is
+    # very slow as we process each file sequentially and does the line length
+    # checking in bash.
+    #
+    # Depending on whether we are in pre-commit mode of full mode, we want to
+    # invoke these scripts differently:
+    # TODO: Consider tidying up this and perhaps solve #1371 in the process.
```

Comment:
Annoyingly, this issue reference is helpfully linkified by GitHub into a dead link, since this actually refers to the `reboot-dev/respect` repo. Simplest workaround:
```suggestion
    # TODO: Consider tidying up this and perhaps solve the above ISSUE in the process.
```